src/bank.rs

```diff
@@ -335,91 +327,130 @@ impl Bank {
     }
 
     pub fn verify_transaction(
-        tx: &Transaction,
+        pix: usize,
+        tx_program_id: &Pubkey,
         pre_program_id: &Pubkey,
         pre_tokens: i64,
         account: &Account,
     ) -> Result<()> {
         // Verify the transaction
         // make sure that program_id is still the same or this was just assigned by the system call contract
         if !((*pre_program_id == account.program_id)
-            || (SystemProgram::check_id(&tx.program_id)
+            || (SystemProgram::check_id(&tx_program_id)
                 && SystemProgram::check_id(&pre_program_id)))
         {
             //TODO, this maybe redundant bpf should be able to guarantee this property
-            return Err(BankError::ModifiedContractId);
+            return Err(BankError::ModifiedContractId(pix as u8));
         }
         // For accounts unassigned to the contract, the individual balance of each accounts cannot decrease.
-        if tx.program_id != account.program_id && pre_tokens > account.tokens {
-            return Err(BankError::ExternalAccountTokenSpend);
+        if *tx_program_id != account.program_id && pre_tokens > account.tokens {
+            return Err(BankError::ExternalAccountTokenSpend(pix as u8));
         }
         if account.tokens < 0 {
-            return Err(BankError::ResultWithNegativeTokens);
+            return Err(BankError::ResultWithNegativeTokens(pix as u8));
         }
         Ok(())
     }
 
-    fn loaded_contract(&self, tx: &Transaction, accounts: &mut [Account]) -> bool {
+    fn loaded_contract(
+        &self,
+        tx_program_id: &Pubkey,
+        tx: &Transaction,
+        pix: usize,
+        accounts: &mut [&mut Account],
+    ) -> bool {
         let loaded_contracts = self.loaded_contracts.write().unwrap();
-        match loaded_contracts.get(&tx.program_id) {
+        match loaded_contracts.get(&tx_program_id) {
             Some(dc) => {
-                let mut infos: Vec<_> = (&tx.keys)
+                let mut infos: Vec<_> = (&tx.account_keys)
                     .into_iter()
                     .zip(accounts)
                     .map(|(key, account)| KeyedAccount { key, account })
                     .collect();
 
-                dc.call(&mut infos, &tx.userdata);
+                dc.call(&mut infos, tx.userdata(pix));
                 true
             }
             None => false,
         }
     }
-
+    fn with_subset<F, A>(accounts: &mut [Account], ixes: &[u8], func: F) -> A
```

Comment:
This looks like you lost a fight with the borrow-checker. Can you add a long code comment for why you wouldn't return the subset `Vec<&mut Account>` to the caller? I could see how that'd be challenging, but if so, we'd want to at least create a ticket to go clean this up later.

Reply:
@garious, you can't move the references around like that.  Basically, the borrow checker wont let you take a "write" lock twice on the same ref, and it has no way to understand how the sort/reorder maintains that assertion.

Reply:
@garious added comments to the function